sky/serve/spot_policy.py

```diff
@@ -0,0 +1,132 @@
+    NAME: Optional[str] = None
+    REGISTRY: Dict[str, Type['SpotPlacer']] = dict()
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        assert spec.spot_zones is not None
+        self.zones = list(spec.spot_zones)
+        self.zone2type: Dict[str, SpotZoneType] = {
+            zone: SpotZoneType.ACTIVE for zone in self.zones
+        }
+
+    def __init_subclass__(cls) -> None:
+        if cls.NAME is None:
+            # This is an abstract class, don't put it in the registry.
+            return
+        assert cls.NAME not in cls.REGISTRY, f'Name {cls.NAME} already exists'
+        cls.REGISTRY[cls.NAME] = cls
+
+    @classmethod
+    def get_policy_names(cls) -> List[str]:
+        return list(cls.REGISTRY.keys())
+
+    def select(self, existing_replicas: List['replica_managers.ReplicaInfo'],
+               num_zones: int) -> List[str]:
+        """Select next zone to place spot instance."""
+        raise NotImplementedError
+
+    def move_zone_to_active(self, zone: str) -> None:
+        assert zone in self.zone2type
+        self.zone2type[zone] = SpotZoneType.ACTIVE
+
+    def move_zone_to_preempted(self, zone: str) -> None:
+        assert zone in self.zone2type
+        self.zone2type[zone] = SpotZoneType.PREEMPTED
+
+    def handle_active(self, zone: str) -> None:
+        self.move_zone_to_active(zone)
+
+    def handle_preemption(self, zone: str) -> None:
+        self.move_zone_to_preempted(zone)
+
+    def clear_preempted_zones(self) -> None:
+        for zone in self.zones:
+            self.move_zone_to_active(zone)
+
+    def active_zones(self) -> List[str]:
+        return [
+            zone for zone, zone_type in self.zone2type.items()
+            if zone_type == SpotZoneType.ACTIVE
+        ]
+
+    def preempted_zones(self) -> List[str]:
+        return [
+            zone for zone, zone_type in self.zone2type.items()
+            if zone_type == SpotZoneType.PREEMPTED
+        ]
+
+    def __repr__(self) -> str:
+        return f'{self.NAME}SpotPlacer()'
+
+    @classmethod
+    def from_spec(cls, spec: 'service_spec.SkyServiceSpec') -> 'SpotPlacer':
+        assert (spec.spot_placer is not None and
+                spec.spot_placer in cls.REGISTRY)
+        return cls.REGISTRY[spec.spot_placer](spec)
+
+
+class DynamicFailoverSpotPlacer(SpotPlacer):
+    """Dynamic failover to an active zone when preempted."""
+    NAME: Optional[str] = 'DynamicFailover'
+
+    def _filter_unvisited_active_zones(
+            self, existing_replicas: List['replica_managers.ReplicaInfo']
+    ) -> List[str]:
+        existing_zones = set()
+        info = None
+        for info in existing_replicas:
+            if not info.is_spot:
+                # filter on demand fallbacks
+                continue
+            if info.zone is not None:
+                existing_zones.add(info.zone)
+            else:
+                logger.error(f'Zone is None for replica {info}')
```

Comment:
```suggestion
        existing_zones = {info.zone for info in existing_replicas if not info.is_spot and info.zone is not None}
```
nit; also we could inline this function?